src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,419 @@
+        emit AssetsAdded(gameId, assetIds, values);
+    }
+
+    /// @notice Function to remove assets from a GAME
+    /// @param gameId The GAME to remove assets from
+    /// @param assetIds An array of asset Ids to remove
+    /// @param values An array of the number of each asset id to remove
+    /// @param to The address to send removed assets to
+    /// @param uri The URI string to update the GAME token's URI
+
+    function removeAssets(
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values,
+        address to,
+        string memory uri
+    ) public override gameManagerOnly() notToZero(to) {
+        require(assetIds.length == values.length && assetIds.length != 0, "INVALID_INPUT_LENGTHS");
+
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            uint256 currentValue = _gameAssets[gameId][assetIds[i]];
+            require(currentValue != 0 && values[i] != 0 && values[i] <= currentValue, "INVALID_ASSET_REMOVAL");
+            _gameAssets[gameId][assetIds[i]] = currentValue.sub(values[i]);
+        }
+
+        if (assetIds.length == 1) {
+            _asset.safeTransferFrom(address(this), to, assetIds[0], values[0], "");
+        } else {
+            _asset.safeBatchTransferFrom(address(this), to, assetIds, values, "");
+        }
+
+        _setTokenURI(gameId, uri);
+        emit AssetsRemoved(gameId, assetIds, values, to);
+    }
+
+    /// @notice Get the creator of the token type `id`.
+    /// @param id the id of the token to get the creator of.
+    /// @return the creator of the token type `id`.
+    function creatorOf(uint256 id) public view override returns (address) {
+        require(id != uint256(0), "GAME_NEVER_MINTED");
+        address originalCreator = address(id / CREATOR_OFFSET_MULTIPLIER);
+        address newCreator = _creatorship[originalCreator];
+        if (newCreator != address(0)) {
+            return newCreator;
+        }
+        return originalCreator;
+    }
+
+    /// @notice Return the URI of a specific token
+    /// @param gameId The id of the token
+    /// @return uri The URI of the token
+    function tokenURI(uint256 gameId) public view override returns (string memory uri) {
+        require(_ownerOf(gameId) != address(0), "BURNED_OR_NEVER_MINTED");
+        string memory URI = _metaData[gameId];
+        return URI;
+    }
+
+    /// @notice Function to transfer assets from a burnt GAME
+    /// @param from Previous owner of the burnt game
+    /// @param to Address that will receive the assets
+    /// @param gameId Id of the burnt GAME token
+    /// @param assetIds The assets to recover from the burnt GAME
+    /// @param values The amount of each asset to recover
+    function recoverAssets(
+        address from,
+        address to,
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values
+    ) public override gameManagerOnly() {
+        _recoverAssets(from, to, gameId, assetIds, values);
```

Comment:
I think the recovering of asset should not be manager protected.
if a Game token is burnt, a user should always be able to retrieve its assets.

So I would also add metatx support here

Reply:
@wighawag along this same line of thinking...
What about the `destroyGame` and `destroyAndRecover` functions? 
I feel that perhaps none of these should be manager-protected.
Thoughts?

Reply:
Good questions, I ll ask Buche